core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -462,20 +465,22 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
 
       if (!inPlaceAssignment) {
         // Cannot do in place assignment.
-        val wrapperMessageTimestamp = {
+        val (wrapperMessageTimestamp, offsetOfMaxTimestampInMessageSet) = {
           if (messageFormatVersion == Message.MagicValue_V0)
-            Some(Message.NoTimestamp)
-          else if (messageFormatVersion > Message.MagicValue_V0 && messageTimestampType == TimestampType.CREATE_TIME)
-            Some(maxTimestamp)
+            (Some(Message.NoTimestamp), -1L)
+          else if (messageTimestampType == TimestampType.CREATE_TIME)
+            (Some(maxTimestamp), {if (targetCodec == NoCompressionCodec) offsetOfMaxTimestamp else offsetCounter.value + validatedMessages.length - 1})
           else // Log append time
-            Some(now)
+            (Some(now), {if (targetCodec == NoCompressionCodec) offsetCounter.value else offsetCounter.value + validatedMessages.length - 1})
         }
 
-        (new ByteBufferMessageSet(compressionCodec = targetCodec,
-                                  offsetCounter = offsetCounter,
-                                  wrapperMessageTimestamp = wrapperMessageTimestamp,
-                                  timestampType = messageTimestampType,
-                                  messages = validatedMessages: _*), true)
+        ValidationAndOffsetAssignResult(new ByteBufferMessageSet(compressionCodec = targetCodec,
+                                                                 offsetCounter = offsetCounter,
+                                                                 wrapperMessageTimestamp = wrapperMessageTimestamp,
+                                                                 timestampType = messageTimestampType,
+                                                                 messages = validatedMessages: _*),
+                                        offsetOfMaxTimestampInMessageSet,
+                                        wrapperMessageTimestamp.get, messageSizeMaybeChanged = true)
```

Comment:
It seems that the ordering of the 2nd and the 3rd params is reversed? To avoid this, perhaps it will be clearer to use named parameters since timestamp and offset are of the same type.